### PR TITLE
feat: integrate reduce runtimes

### PR DESCRIPTION
**Pull request by alaeddine-13**
closes: https://github.com/jina-ai/jina/issues/4015

### CODE REVIEW

**Comment by JoanFM** [comment]
@alaeddine-13 please do not merge the addition in the types with the executor part. They shouuld be independent commits, because the second one will not be relevant for `star-routing`

**Review by JoanFM — CHANGES_REQUESTED** [review]
Please, first merge the `types` part with no `Executor` involved

File: tests/integration/reduce_shards/test_reduce.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                assert f'c-{shard}-{chunk}' in chunks
+
+
+@pytest.mark.parametrize('n_shards', [3, 5])
+@pytest.mark.parametrize('n_docs', [3, 5])
+def test_uses_after_no_reduce(n_shards, n_docs):
+    search_flow = Flow().add(
+        uses=MyExecutor,
+        shards=n_shards,
+        uses_after='Executor',
```
Comment (cristianmtr) [inline tests/integration/reduce_shards/test_reduce.py:60]: How can the results be the same with uses_after='Executor' as they are with the Reducer Executor?

**Review by cristianmtr — CHANGES_REQUESTED** [review]

File: tests/integration/reduce_shards/test_reduce.py (hunk 1 above)
Comment (alaeddine-13) [reply tests/integration/reduce_shards/test_reduce.py:60]: It shouldn't be, that's a WIP code that I didn't finish. I'll fix this and properly assert that reduce wasn't executed

**Comment by alaeddine-13** [comment]
> @alaeddine-13 please do not merge the addition in the types with the executor part. They shouuld be independent commits, because the second one will not be relevant for `star-routing`

should be okay now

File: jina/peapods/runtimes/request_handlers/data_request_handler.py [hunk 2]
```diff
@@ -22,7 +23,9 @@ def _get_docs_matrix_from_message(
     field: str,
 ) -> List['DocumentArray']:
     if partial_request is not None:
-        result = [getattr(r, field) for r in reversed(partial_request)]
+        result = [(r.pod_name, getattr(r, field)) for r in reversed(partial_request)]
+        result.sort(key=itemgetter(0))
```
Comment (JoanFM) [inline jina/peapods/runtimes/request_handlers/data_request_handler.py:27]: why is this needed?

File: jina/proto/jina.proto [hunk 3]
```diff
@@ -314,6 +314,7 @@ message RequestProto {
     message DataRequestProto {
         repeated DocumentProto docs = 1; // a list of Documents to query
         repeated DocumentProto groundtruths = 2; // a list of groundtruth Document you want to evaluate it with
+        string pod_name = 3; // sender pod name
```
Comment (JoanFM) [inline jina/proto/jina.proto:317]: do not add this here, why is this needed?

**Review by JoanFM — CHANGES_REQUESTED** [review]
Do not change protobuf specifciation for this

File: jina/peapods/runtimes/request_handlers/data_request_handler.py (hunk 2 above)
Comment (alaeddine-13) [reply jina/peapods/runtimes/request_handlers/data_request_handler.py:27]: This is important to have deterministic results after reducing
It turns out that the document arrays coming from different shards do not have a deterministic order: the one that completes first will be returned first.
As reduce's result (and also match merger) depend on the order of the DocumentArrays, there has to be a way to keep the order deterministic.
I thought about using the pod_name to do so. It's also easy to understand for the user

File: jina/proto/jina.proto (hunk 3 above)
Comment (alaeddine-13) [reply jina/proto/jina.proto:317]: https://github.com/jina-ai/jina/pull/4030#discussion_r767439206

File: jina/peapods/runtimes/request_handlers/data_request_handler.py (hunk 2 above)
Comment (JoanFM) [reply jina/peapods/runtimes/request_handlers/data_request_handler.py:27]: different shards share pod name from the user point of view. This is a weak reason to alter the proto structure so heavily

Comment (alaeddine-13) [reply jina/peapods/runtimes/request_handlers/data_request_handler.py]: Removed.
Results still depend on the order and are undeterministic. Should be solved in star routing

File: jina/proto/jina.proto (hunk 3 above)
Comment (alaeddine-13) [reply jina/proto/jina.proto]: removed

**Review by JoanFM — CHANGES_REQUESTED** [review]
We need to add tests that this work in K8s

File: jina/peapods/runtimes/request_handlers/data_request_handler.py [hunk 4]
```diff
@@ -121,7 +128,7 @@ def handle(
             msg.envelope.header.exec_endpoint not in self._executor.requests
             and __default_endpoint__ not in self._executor.requests
         ):
-            self.logger.debug(
+            self.logger.info(
```
Comment (JoanFM) [inline jina/peapods/runtimes/request_handlers/data_request_handler.py:131]:
```suggestion
            self.logger.debug(
```

**Review by JoanFM — CHANGES_REQUESTED** [review]
Small thing

**Review by JoanFM — APPROVED** [review]
